File: week8/task_C.cpp
#include <iostream>
 
struct Node {
    int key, height, size;
    Node* left;
    Node* right;
    Node(int key): key(key), height(1), size(0),
                   left(nullptr), right(nullptr) {}
};
 
int size(Node* p) {
    return !p ? -1 : p->size;
}
 
void correct_size(Node* p) {
    p->size = size(p->left) + size(p->right) + 2;
}
 
int height(Node* p) {
    return !p ? 0 : p->height;
}
 
void correct_height(Node* p) {
    p->height = (height(p->left) > height(p->right) ? height(p->left) : height(p->right)) + 1;
}
 
int balance_differents(Node* p) {
    return height(p->right) - height(p->left);
}
 
Node* rotate_left(Node* p) {
    Node* right = p->right;
    p->right = right->left;
    right->left = p;
    correct_size(p);
    correct_size(right);
    correct_height(p);
    correct_height(right);
    return right;
}
 
Node* rotate_right(Node* p) {
    Node* left = p->left;
    p->left = left->right;
    left->right = p;
    correct_size(p);
    correct_size(left);
    correct_height(p);
    correct_height(left);
    return left;
}
 
Node* balance(Node* p) {
    correct_size(p); 
    correct_height(p);
    if (balance_differents(p) == 2) {
        if (balance_differents(p->right) < 0) {
            p->right = rotate_right(p->right);
        }
        return rotate_left(p);
    }
    if (balance_differents(p) == -2) {
        if (balance_differents(p->left) > 0) {
            p->left = rotate_left(p->left);
        }
        return rotate_right(p);
    }
    return p;
}
 
Node* find_k(Node* p, int k) {
    if (!p->right) {
        return (k == 1) ? p : find_k(p->left, k - 1);
    }
    if (k == 2 + size(p->right)) {
        return p;
    }
    if (k > 2 + size(p->right)) {
        return find_k(p->left, k - 2 - size(p->right));
    }
    return find_k(p->right, k);
}
 
Node* insert(Node* p, int key) {
    if (!p) {
        return new Node(key);
    }
    if (p->key > key) {
        p->left = insert(p->left, key);
    } else {
        p->right = insert(p->right, key);
    }
    correct_size(p);
    correct_height(p);
    return balance(p);
}
 
Node* find_min(Node* p) {
    return !p->left ? p : find_min(p->left);
}
 
Node* remove_min(Node* p) {
    if (!p->left) {
        return p->right;
    }
    p->left = remove_min(p->left);
    correct_size(p);
    correct_height(p);
    return balance(p);
}
 
Node* remove(Node* p, int key) {
    if (!p) {
        return nullptr;
    }
    if (key < p->key) {
        p->left = remove(p->left, key);
        if (p->left) {
            correct_size(p->left);
            correct_height(p->left);
        }
    } else if (key > p->key) {
        p->right = remove(p->right, key);
        if (p->right) {
            correct_height(p->right);
            correct_size(p->right);
        }
    } else {
        Node* left_child = p->left;
        Node* right_child = p->right;
        if (!right_child) {
            return left_child;
        }
        Node* min = find_min(right_child);
        min->right = remove_min(right_child);
        min->left = left_child;
        correct_size(min);
        correct_height(min);
        return balance(min);
    }
    return balance(p);
}
 
int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
 
    int n, c, k;
    std::cin >> n;
    Node *tree = nullptr;
    for (int i = 0; i < n; ++i) {
        std::cin >> c >> k;
        if (c == 1) {
            tree = insert(tree, k);
        } else if (c == 0) {
            std::cout << find_k(tree, k)->key << std::endl;
        } else if (c == -1) {
            tree = remove(tree, k);
        }
    }
 
    return 0;
}

File: week8/python/task_C.py
EMPTY_LINK = None
EMPTY_KEY = None
# OPERATIONS CODES
INSERT_OP = 1
DELETE_OP = -1
FIND_K_OP = 0
 
 
class NodeAVL:
 
    def __init__(self, key=EMPTY_KEY, parent=EMPTY_LINK):
        self.key = key
        self.parent = parent
        self.height = 1
        self.size = 1
        
        self.left = EMPTY_LINK
        self.right = EMPTY_LINK
    
    def node_params(self, node):
        if not node == EMPTY_LINK:  # if not empty!
            return node.size, node.height
        return 0, 0
    
    def subtrees_heights(self):
        '''
        Returns heights and sizes of the node's subtrees
            Return format: int left_size, 
                           int height_left, 
                           int size_right, 
                           int height_right
        '''
        size_left, height_left = self.node_params(self.left)
        size_right, height_right = self.node_params(self.right) 
        return size_left, height_left, size_right, height_right
    
    def balance_factor(self):
        '''
        Returns a balance factor of the Node object (self)
            Return format: int
        '''
        _, height_left, _, height_right = self.subtrees_heights()
        return height_right - height_left
    
    def height_update(self):
        '''
        Updates the Node object (self) height
        '''
        size_left, height_left, size_right, height_right = self.subtrees_heights()
        self.size = size_left + size_right + 1
        if height_left > height_right:
            self.height = height_left + 1
        else:
            self.height = height_right + 1
            
    # ROTATIONS
    def rotate_right(self, node):
        '''
        Perform a right rotation around the Node object (self)
        '''
        # update links
        left_node = node.left
        node.left = left_node.right
        left_node.right = node
        # update parents
        left_node.parent = node.parent
        node.parent = left_node
        # update heights
        node.height_update()
        left_node.height_update()
        
        return left_node
    
    def rotate_left(self, node):
        '''
        Perform a left rotation around the Node object (self)
        '''
        # update links
        right_node = node.right
        node.right = right_node.left
        right_node.left = node
        # update parents
        right_node.parent = node.parent
        node.parent = right_node
        # update heights
        node.height_update()
        right_node.height_update()
        
        return right_node
    
    def balance(self, node):
        node.height_update()
        if node.balance_factor() == 2:
            if node.right.balance_factor() < 0:
                node.right = self.rotate_right(node.right)
            return self.rotate_left(node)
        if node.balance_factor() == -2:
            if node.left.balance_factor() > 0:
                node.left = self.rotate_left(node.left)
            return self.rotate_right(node)
        return node  # no balance needed
    
    # OPERATIONS
    def insert(self, node, key, parent=EMPTY_LINK):
        if node == EMPTY_LINK or node.key == EMPTY_KEY:
            return NodeAVL(key=key, parent=parent)  # add new Node()
        if key < node.key:
            node.left = self.insert(node.left, key, parent=node)
        if key > node.key:
            node.right = self.insert(node.right, key, parent=node)
        return self.balance(node)
    
    def delete(self, node, key):
        if node == EMPTY_LINK or node.key == EMPTY_KEY:
            return node
        if key < node.key:
            node.left = self.delete(node.left, key)
        elif key > node.key:
            node.right = self.delete(node.right, key)
        else:
            if node.left != EMPTY_LINK and node.right != EMPTY_LINK:  # two childs
                node.key = self.find_max(node.left).key
                node.left = self.delete(node.left, node.key)  # delete max
            else:  # one or no childs!
                if node.left == EMPTY_LINK:
                    if node.right == EMPTY_LINK:  # no childs
                        if node.parent == EMPTY_LINK:  # root is the only element in tree!
                            node.key = EMPTY_KEY
                        else:
                            return EMPTY_LINK
                    else:  # one child: node.right != EMPTY_LINK
                        node.right.parent = node.parent
                        node = node.right
                else:  # one child: node.left != EMPTY_LINK
                    node.left.parent = node.parent
                    node = node.left
        return self.balance(node)
 
    def find_max(self, node):
        while not node.right == EMPTY_LINK:
            node = self.find_max(node.right)
        return node
    
    def find_k_max(self, node, k):
        if node.left == EMPTY_LINK:
            this_max_ind = node.size
        else:
            this_max_ind = node.size - node.left.size
        if this_max_ind == k:
            return node
        else:
            if this_max_ind > k:
                return self.find_k_max(node.right, k)
            else:
                return self.find_k_max(node.left, k - this_max_ind)
 
 
tree = NodeAVL()
n_operations = int(input())
for _ in range(n_operations):
    op_i, k_i = map(int, input().split())
    if op_i == INSERT_OP:
        tree = tree.insert(tree, k_i)
    elif op_i == DELETE_OP:
        tree = tree.delete(tree, k_i)
    else:  # find k'th
        print(tree.find_k_max(tree, k_i).key)
